### PR TITLE
feat: Add hpa on compute node

### PR DESCRIPTION
**Pull request by lianneli**
add hpa on compute node

fix issue #72

### CODE REVIEW

**Comment by tcodehuber** [comment]
LGTM, thanks for your contribution!

**Comment by lianneli** [comment]
Tested on local env.

image

image

File: charts/risingwave/templates/autoscaler.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    {{- tpl (toYaml .Values.commonLabels) . | nindent 4 }}
+  {{- end }}
+spec:
+  maxReplicas: {{ .Values.computeComponent.autoscaling.maxReplicas }}
+  minReplicas: {{ .Values.computeComponent.replicas }}
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: StatefulSet
+    name: {{ include "risingwave.computeComponentName" . }}
+  {{ if .Values.computeComponent.autoscaling.metrics -}}
```
Comment (arkbriar) [inline charts/risingwave/templates/autoscaler.yaml:23]: nit: prefer 
```suggestion
  {{- if .Values.computeComponent.autoscaling.metrics }}
```

File: charts/risingwave/templates/autoscaler.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  metrics:
+    {{- toYaml .Values.computeComponent.autoscaling.metrics | nindent 4}}
+  {{- end }}
+  {{ if .Values.computeComponent.autoscaling.behavior -}}
```
Comment (arkbriar) [inline charts/risingwave/templates/autoscaler.yaml:27]: ditto
```suggestion
  {{- if .Values.computeComponent.autoscaling.behavior }}
```

File: charts/risingwave/values.yaml [hunk 3]
```diff
@@ -971,6 +971,13 @@ computeComponent:
   ##
   replicas: 1
 
+  ## @param computeComponent.autoscaling Compute Pod horizontal Pod Autoscalers
```
Comment (arkbriar) [inline charts/risingwave/values.yaml:974]: Add a reference to the details of the autoscaler object.

```suggestion
  ## @param computeComponent.autoscaling Compute Pod horizontal Pod Autoscalers
  ## Reference: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#container-resource-metrics
  ##
```

File: charts/risingwave/templates/autoscaler.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+{{- if .Values.computeComponent.autoscaling.enabled -}}
+apiVersion: autoscaling/v2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: {{ include "risingwave.computeComponentName" . }}-autoscaler
+  namespace: {{ .Release.Namespace }}
+  {{- if .Values.commonLabels }}
+  labels:
+    {{- tpl (toYaml .Values.commonLabels) . | nindent 4 }}
+  {{- end }}
```
Comment (arkbriar) [inline charts/risingwave/templates/autoscaler.yaml:15]: Missing annotations.

**Review by arkbriar — COMMENTED** [review]
Mostly LGTM! Thanks!

Besides, could you help add tests cases for the auto scaler? Expected behaviors should be:
- when it's disabled, `autoscaler.yaml` contains no object
- when it's enabled, `autoscaler.yaml` contains a `HorizontalPodAutoscaler`, and it reflects the values.

**Comment by arkbriar** [comment]
Besides, the repo requires all commits to be signed to merge the PR. Please follow the [Learn more about signing commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) and sign your commits. You may need to force update the PR after that. :)

File: charts/risingwave-operator/templates/role.yaml
[the remark this answers is not in the document]
Comment (lianneli) [reply charts/risingwave-operator/templates/role.yaml:1]: fixed.

File: charts/risingwave/templates/autoscaler.yaml (hunk 1 above)
Comment (lianneli) [reply charts/risingwave/templates/autoscaler.yaml]: fixed.

File: charts/risingwave/templates/autoscaler.yaml (hunk 2 above)
Comment (lianneli) [reply charts/risingwave/templates/autoscaler.yaml]: fixed.

File: charts/risingwave/values.yaml (hunk 3 above)
Comment (lianneli) [reply charts/risingwave/values.yaml:974]: fixed.

**Comment by lianneli** [comment]
@arkbriar Thanks for you advice. Could you please review again?

File: charts/risingwave/templates/autoscaler.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  {{- if .Values.commonAnnotations }}
+  annotations:
+    {{- tpl (toYaml .Values.commonAnnotations) . | nindent 4 }}
+  {{- end }}
```
Comment (arkbriar) [inline charts/risingwave/templates/autoscaler.yaml:19]: Forgot to mention, there are two helper functions to add the labels and annotations. Prefer using them instead.

```suggestion
  labels:
    {{- include "risingwave.labels" . | nindent 4 }}
    risingwave.risingwavelabs.com/component: compute
  {{- $annotations := (include "risingwave.annotations" . ) | trim }}
  {{- if $annotations }}
  annotations:
    {{ nindent 4 $annotations }}
  {{- end }}
```

File: charts/risingwave/tests/autoscaler_test.yaml [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+          path: metadata.annotations
+          content:
+            ANNOTATION: ANNOTATION_V
+  - it: HorizontalPodAutoscaler should reflect computeComponent replicas and autoscaling maxReplicas
+    set:
+      computeComponent:
+        autoscaling:
+          enabled: true
+          maxReplicas : 5
+      replicas: 1
```
Comment (arkbriar) [inline charts/risingwave/tests/autoscaler_test.yaml:52]:
```suggestion
      computeComponent:
        autoscaling:
          enabled: true
          maxReplicas : 5
        replicas: 1
```

File: charts/risingwave/tests/autoscaler_test.yaml [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+          value: 0
+      - contains:
+          path: spec.behavior.scaleUp.policies
+          content:
+            type: Percent
+            value: 100
+            periodSeconds: 15
+      - equal:
+          path: spec.behavior.scaleUp.selectPolicy
+          value: Max
```
Comment (arkbriar) [inline charts/risingwave/tests/autoscaler_test.yaml:116]: nit: prefer an empty line at EOF. Run `make sanitize` to check and add to all files.

**Review by arkbriar — COMMENTED** [review]
Others LGTM!

File: charts/risingwave/tests/autoscaler_test.yaml (hunk 7 above)
Comment (lianneli) [reply charts/risingwave/tests/autoscaler_test.yaml]: got and fixed 😄

**Comment by lianneli** [comment]
@arkbriar Thanks again and I have accepted all of them.  May be other suggestions?

**Review by arkbriar — APPROVED** [review]
LGTM!

**Pull request merged**